bigtable/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/core/CallProxy.java

```diff
@@ -0,0 +1,169 @@
+    // customer's client is ready.
+    clientCallListener = new ResponseProxy(serverCall);
+  }
+
+  /**
+   * Back pressure aware message pump of request messages from a customer's downstream client to
+   * upstream Bigtable service.
+   *
+   * <p>Additional messages are requested from the downstream while the upstream's isReady() flag is
+   * set. As soon as the upstream signals that is full by returning false for isReady(). {@link
+   * RequestProxy} will remember that the need to get more messages from downstream and then wait
+   * until the upstream signals readiness via onClientReady().
+   *
+   * <p>Please note in the current Bigtable protocol, all RPCs a client unary. Until that changes,
+   * this proxy will only have a single iteration. However, its designed generically to support
+   * future usecases.
+   */
+  private class RequestProxy extends ServerCall.Listener<ReqT> {
+
+    private final ClientCall<ReqT, ?> clientCall;
+
+    @GuardedBy("this")
+    private boolean needToRequest;
+
+    public RequestProxy(ClientCall<ReqT, ?> clientCall) {
+      this.clientCall = clientCall;
+    }
+
+    @Override
+    public void onCancel() {
+      clientCall.cancel("Server cancelled", null);
+    }
+
+    @Override
+    public void onHalfClose() {
+      clientCall.halfClose();
+    }
+
+    @Override
+    public void onMessage(ReqT message) {
+      clientCall.sendMessage(message);
+      synchronized (this) {
+        if (clientCall.isReady()) {
+          clientCallListener.serverCall.request(1);
+        } else {
+          // The outgoing call is not ready for more requests. Stop requesting additional data and
+          // wait for it to catch up.
+          needToRequest = true;
+        }
+      }
+    }
+
+    @Override
+    public void onReady() {
+      clientCallListener.onServerReady();
+    }
+
+    // Called from ResponseProxy, which is a different thread than the ServerCall.Listener
+    // callbacks.
+    synchronized void onClientReady() {
+      if (needToRequest) {
+        // When the upstream client is ready for another request message from the customer's client,
+        // ask for one more message.
+        clientCallListener.serverCall.request(1);
+        needToRequest = false;
+      }
+    }
+  }
+
+  /**
+   * Back pressure aware message pump of response messages from upstream Bigtable service to a
+   * customer's downstream client.
+   *
+   * <p>Additional messages are requested from the upstream while the downstream's isReady() flag is
+   * set. As soon as the downstream signals that is full by returning false for isReady(). {@link
+   * ResponseProxy} will remember that the need to get more messages from upstream and then wait
+   * until the downstream signals readiness via onServerReady().
+   */
+  private class ResponseProxy extends ClientCall.Listener<RespT> {
+
+    private final ServerCall<?, RespT> serverCall;
+    // Hold 'this' lock when accessing
```

Comment:
Nit: Should we replace the comment with GuardedBy annotation?

Reply:
done